### ISSUE TITLE
Model trains for more steps than requested

### ISSUE THREAD

**Issue by sayakpaul** [a-r-r-o-w/finetrainers#216]
I am doing a full finetuning run of CogVideoX-5B and requested for 2000 `train_steps` All arguments are here: https://wandb.ai/sayakpaul/finetrainers-ltxv/runs/snv2aphi/overview. 

As we can see here training continues for MORE than 2k steps: https://wandb.ai/sayakpaul/finetrainers-ltxv/runs/snv2aphi/workspace. 

Is this expected?

**Comment by sayakpaul** [comment]
Smaller command to reproduce:

<details>
<summary>Command</summary>

```bash
export NCCL_P2P_DISABLE=1
export TORCH_NCCL_ENABLE_MONITORING=0
export NCCL_ASYNC_ERROR_HANDLING=1
export FINETRAINERS_LOG_LEVEL=DEBUG

GPU_IDS="0,1,2,3,4,5,6,7"

DATA_ROOT="video-dataset-disney"
CAPTION_COLUMN="prompt.txt"
VIDEO_COLUMN="videos.txt"
OUTPUT_DIR="cogvideox-disney"

# Model arguments
model_cmd="--model_name cogvideox \
  --pretrained_model_name_or_path THUDM/CogVideoX-5b"

# Dataset arguments
ID_TOKEN="DISNEY_FX"
dataset_cmd="--data_root $DATA_ROOT \
  --video_column $VIDEO_COLUMN \
  --caption_column $CAPTION_COLUMN \
  --id_token $ID_TOKEN \
  --precompute_conditions \
  --video_resolution_buckets 49x512x768 \
  --caption_dropout_p 0.05"

# Dataloader arguments
dataloader_cmd="--dataloader_num_workers 0"

# Diffusion arguments
diffusion_cmd=""

# Training arguments
training_cmd="--training_type full-finetune \
  --seed 42 \
  --batch_size 1 \
  --train_steps 10 \
  --gradient_accumulation_steps 4 \
  --gradient_checkpointing \
  --checkpointing_steps 2 \
  --checkpointing_limit 2 \
  --enable_slicing \
  --enable_tiling"

# Optimizer arguments
optimizer_cmd="--optimizer adamw \
  --lr 5e-6 \
  --lr_scheduler constant_with_warmup \
  --lr_warmup_steps 200 \
  --lr_num_cycles 1 \
  --beta1 0.9 \
  --beta2 0.95 \
  --weight_decay 1e-2 \
  --epsilon 1e-8 \
  --max_grad_norm 1.0"

# Validation arguments
validation_cmd="--validation_prompts \"$ID_TOKEN The video shows a news segment from a Hindi news channel. The first shot is of a male news anchor in a suit and glasses speaking directly to the camera. A lower third graphic displays text in Hindi along with “Breaking News” in English. The date and time are shown in the top left corner of the screen. The next shot is a graphic transition with orange and red tones, resembling fire and destruction. The following scenes show a series of transitions, again with fire-like imagery, and text overlays in Hindi that translate to “Big Bulletin.” The words “Coming Up” are also visible in English. The overall style is dramatic, suggestive of breaking news coverage. The Hindi language and visual style suggest the newscast originates from India.@@@49x512x768\" \
  --num_validation_videos 1 \
  --validation_steps 2"

# Miscellaneous arguments
miscellaneous_cmd="--tracker_name finetrainers-cogvideox \
  --output_dir $OUTPUT_DIR \
  --nccl_timeout 1800 \
  --report_to wandb"

cmd="accelerate launch --config_file accelerate_configs/deepspeed.yaml --gpu_ids $GPU_IDS train.py \
  $model_cmd \
  $dataset_cmd \
  $dataloader_cmd \
  $diffusion_cmd \
  $training_cmd \
  $optimizer_cmd \
  $validation_cmd \
  $miscellaneous_cmd"

echo "Running command: $cmd"
eval $cmd
echo -ne "-------------------- Finished executing script --------------------\n\n"
```

</details>

**Comment by sayakpaul** [comment]
@a-r-r-o-w a gentle ping.

**Comment by a-r-r-o-w** [comment]
This is honestly very weird, not sure how or why it happens.

Will be fixing in #245. With that PR, the single source of truth of the number of images/videos seen by the model will be the number of training steps.

Currently, if we use 2000 training steps for example:
- on 1 GPU, it is 2000 effective videos seen
- but on 8 GPU, it is 16000 effective videos seen

Similar for gradient accumulation.

This behaviour is confusing and makes ablation experiments on batch size harder and more time consuming (hard to keep in mind all the parameters that need to be adjusted to keep effective videos constant). With the implementation I have locally for #245, using 2000 training steps will ensure that irrespective of number of GPUs and gradient accumulation steps, the total number of images/videos seen by the model will be 2000, which seems like a more sensible choice and less prone to any bugs from rounding in calculations of effective batch size.

**Comment by a-r-r-o-w** [comment]
Will also help avoid confusion like here about total time required for experiments: https://github.com/a-r-r-o-w/finetrainers/issues/111#issuecomment-2516222917

**Comment by sayakpaul** [comment]
Thank you sir! This is very nice.

**Comment by a-r-r-o-w** [comment]
Addressed now so I believe we can close this. The following takes into account the number of training steps and max number of data points you want the model to see. It assumes that the data comes from an infinite buffer (which is true due to how the dataset are initialized) and handles all the preprocessing & precomputation somewhat efficiently (there are lots of potential improvements possible but it's definitely at a state better than before)

https://github.com/a-r-r-o-w/finetrainers/blob/60ab7f072d7ab61ba489f98fdaee8744980192e1/finetrainers/trainer/sft_trainer/trainer.py#L404-L437
